packages/babylon/src/parser/expression.js

```diff
@@ -417,7 +417,252 @@ export default class ExpressionParser extends LValParser {
+      this.next();
+      const node = this.startNodeAt(startPos, startLoc);
+
+      if (this.match(tt.braceR)) {
+        this.raise(this.state.pos, "there are no clauses in match expression");
+      }
+
+      const firstClause = this.parseMatchClause();
+
+      node.expression = tmp.arguments[0];
+      node.clauses = [firstClause];
+
+      while (this.match(tt.comma)) {
+        this.next();
+
+        if (this.match(tt.braceR)) {
+          break;
+        }
+
+        node.clauses.push(this.parseMatchClause());
+      }
+
+      this.eat(tt.braceR);
+      return this.finishNode(node, "MatchExpression");
+    } else {
+      return tmp;
+    }
+  }
+
+  // pattern '=>' expression
+  parseMatchClause(): N.MatchClause {
+    const node = this.startNode();
+
+    const pattern = this.parseMatchPattern();
+
+    if (this.match(tt._if)) {
+      node.guard = this.parseMatchGuard();
+    }
+
+    if (this.match(tt.eq)) {
+      this.next();
+      node.initializer = this.parseMaybeAssign(false, node.loc.start);
+    }
+
+    if (!this.eat(tt.arrow)) {
+      this.unexpected(this.state.pos, tt.arrow);
+    }
+
+    node.pattern = pattern;
+    this.parseClauseBody(node, true);
+
+    return this.finishNode(node, "MatchClause");
+  }
+
+  parseMatchGuard(): N.MatchGuard {
+    const node = this.startNode();
+
+    if (!this.eat(tt._if)) {
+      this.unexpected(this.state.pos, tt._if);
+    }
+
+    if (!this.eat(tt.parenL)) {
+      this.unexpected(this.state.pos, tt.parenL);
+    }
+
+    node.body = this.parseExpression();
+
+    if (!this.eat(tt.parenR)) {
+      this.unexpected(this.state.pos, tt.parenR);
+    }
+
+    return this.finishNode(node, "MatchGuard");
+  }
+
+  parseClauseBody(node: N.MatchClause, allowExpression: ?boolean): void {
+    const isExpression = allowExpression && !this.match(tt.braceL);
+
+    const oldInParameters = this.state.inParameters;
+    const oldInAsync = this.state.inAsync;
+    this.state.inParameters = false;
+    this.state.inAsync = false;
+
+    if (isExpression) {
+      node.body = this.parseMaybeAssign();
+      node.expression = true;
+    } else {
+      // Start a new scope with regard to labels and the `inGenerator`
+      // flag (restore them to their old value afterwards).
+      const oldInGen = this.state.inGenerator;
+      const oldInFunc = this.state.inFunction;
+      const oldLabels = this.state.labels;
+      this.state.inGenerator = false;
+      this.state.inFunction = true;
+      this.state.labels = [];
+      node.body = this.parseBlock(true);
+      node.expression = false;
+      this.state.inFunction = oldInFunc;
+      this.state.inGenerator = oldInGen;
+      this.state.labels = oldLabels;
+    }
+    this.state.inAsync = oldInAsync;
+
+    this.state.inParameters = oldInParameters;
+  }
+
+  parseMatchPattern(): N.MatchExpressionPattern | null {
+    let node;
+
+    if (this.match(tt.braceL)) {
+      return this.parseObjectMatchPattern();
```

Comment:
I'm just wondering why:

```js
const val = match (res) {
  {'b': s} => true
}
```
```
repl: Unexpected token (4:3)

  2 | 
  3 | const val = match (res) {
> 4 |   {'b': s} => true
    |    ^
  5 | }
```

While: 
```js
const {'b': s} = true;
```

Works fine. From my understanding they are both using this `parseObjectMatchPattern`?

Reply:
Yep, it should be right, it's just not supported currently.

Reply:
At first glance I thought that your code already supported it but you are right. That `parseMatchProperty` is missing the quote handling.